distributed/scheduler.py

```diff
@@ -4245,167 +4249,164 @@ async def add_nanny(self) -> dict[str, Any]:
         }
         return msg
 
-    def update_graph_hlg(
+    def update_graph(
```

Comment:
Everything below is a very large and convoluted change. If there are significant changes in the logic, it might be worth separating those from the refactoring that extracts code into additional methods to make a review easier.

Reply:
Factoring out the code changes themselves is not quite trivial. Particularly the way annotations are handled is something that is highly convoluted between old and new systems so this is effectively a rewrite.

A factored out version of this is basically what's shown in https://github.com/dask/distributed/pull/6028 since it converts annotations back and forth multiple times.

The feedback I'm looking for here is mostly about structure of the new code rather than "do the changes compared to the old code make any sense".

I didn't write any new logic (apart from the annotations) but rather reordered things and encapsulated a couple of operations